embedded-tests/src/test/java/org/apache/druid/testing/embedded/docker/IngestionDockerTest.java

```diff
@@ -0,0 +1,296 @@
+import org.apache.druid.testing.embedded.indexing.Resources;
+import org.apache.druid.testing.embedded.junit5.EmbeddedClusterTestBase;
+import org.apache.druid.testing.embedded.minio.MinIOStorageResource;
+import org.apache.druid.timeline.DataSegment;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.joda.time.DateTime;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ThreadLocalRandom;
+
+/**
+ * Runs some basic ingestion tests using {@link DruidContainers}.
+ */
+public class IngestionDockerTest extends EmbeddedClusterTestBase
+{
+  static {
+    System.setProperty("druid.testing.docker.image", "apache/druid:35.0.0-SNAPSHOT");
+  }
+
+  // Druid Docker containers
+  protected final DruidContainerResource overlordLeader = DruidContainers.newOverlord().withTestImage();
+  protected final DruidContainerResource coordinator = DruidContainers.newCoordinator().withTestImage();
+  protected final DruidContainerResource historical = DruidContainers.newHistorical().withTestImage();
+  protected final DruidContainerResource broker1 = DruidContainers.newBroker().withTestImage();
+  protected final DruidContainerResource middleManager = DruidContainers
+      .newMiddleManager()
+      .withTestImage()
+      .addProperty("druid.segment.handoff.pollDuration", "PT0.1s");
+
+  // Follower EmbeddedOverlord to watch segment publish events
+  private final EmbeddedOverlord overlordFollower = new EmbeddedOverlord()
+      .addProperty("druid.plaintextPort", "7090")
+      .addProperty("druid.manager.segments.useIncrementalCache", "always")
+      .addProperty("druid.manager.segments.pollDuration", "PT0.1s");
+
+  // Additional EmbeddedBroker to wait for segments to become queryable
+  private final EmbeddedBroker broker2 = new EmbeddedBroker()
+      .addProperty("druid.plaintextPort", "7082");
+
+  private final KafkaResource kafkaServer = new KafkaResource();
+
+  @Override
+  public EmbeddedDruidCluster createCluster()
+  {
+    final EmbeddedDruidCluster cluster = EmbeddedDruidCluster.withZookeeper();
+
+    return cluster
+        .useDruidContainers()
+        .useLatchableEmitter()
+        // Needed for overlordFollower to recognize the KafkaSupervisor type
+        .addExtension(KafkaIndexTaskModule.class)
+        .addResource(new EmbeddedDerbyMetadataResource())
+        .addResource(new MinIOStorageResource())
+        .addResource(kafkaServer)
+        .addCommonProperty(
+            "druid.extensions.loadList",
+            "[\"druid-s3-extensions\", \"druid-kafka-indexing-service\", \"druid-multi-stage-query\"]"
+        )
+        .addResource(coordinator)
+        .addResource(overlordLeader)
+        .addResource(middleManager)
+        .addResource(historical)
+        .addResource(broker1)
+        .addServer(overlordFollower)
```

Comment:
a comment about why we mix containers and noncontainers in this test would be helpful.

Reply:
added.